### PR TITLE
bugfix: check if table meta cache should be refreshed in AT mode (2.x)

### PR DESCRIPTION
**Pull request by Bughue**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
同步 https://github.com/seata/seata/pull/4734

### CODE REVIEW

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 1]
```diff
@@ -259,6 +262,75 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
         return records;
     }
 
+    /**
+     * check if the column is null and return
+     *
+     * @param tmeta the table meta
+     * @param colName the column nmae
+     */
+    private static ColumnMeta checkAndGetColumnMeta(TableMeta tmeta , String colName) throws SQLException {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:271]: Method names should be more elegant.

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 2]
```diff
@@ -259,6 +262,75 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
...
+     * @param resultSet the result set
+     * @param statementProxy the statement proxy
+     * @return the table records
+     * @throws SQLException the sql exception
+     */
+    public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet, StatementProxy statementProxy)
+        throws SQLException {
+        try {
+            return buildRecords(tmeta, resultSet);
+        } catch (TableMetaException e) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:293]: TableMetaException should be lightweight and contain only the necessary information.

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 3]
```diff
@@ -259,6 +262,75 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
...
+            // try to build again after refresh table meta successfully
+            return buildRecords(getCacheTableMeta(statementProxy.getConnectionProxy(), tmeta.getTableName()), resultSet);
+        }
+    }
+
+
+    private static void refreshTableMeta(ConnectionProxy connectionProxy, TableMeta tmeta, String columnName)
+        throws SQLException {
+        if (columnEmptyAndRefreshable(connectionProxy, tmeta, columnName)) {
+            synchronized (TableRecords.class) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:307]: Is the lock scope appropriate?

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 4]
```diff
@@ -259,6 +262,75 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
...
+                if (columnEmptyAndRefreshable(connectionProxy, tmeta, columnName)) {
+                    TableMetaCacheFactory.getTableMetaCache(connectionProxy.getDbType()).refresh(
+                        connectionProxy.getTargetConnection(), connectionProxy.getDataSourceProxy().getResourceId());
+                    
+                    // still empty after refreshed
+                    if (getCacheTableMeta(connectionProxy, tmeta.getTableName()).getColumnMeta(columnName) == null) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:313]: Is the API used uniformly?

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java [hunk 5]
```diff
@@ -302,13 +301,9 @@ protected TableMeta getTableMeta() {
      * @return the table meta
      */
     protected TableMeta getTableMeta(String tableName) {
-        if (tableMeta != null) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:305]: What is the mapping between TableMeta and Executor or PS ?

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 1 above)
Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:271]: getColumnMeta() ？

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 2 above)
Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:293]: Do you mean ‘refreshable’ shouldn't be inside the TableMetaException?

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 1 above)
Comment (slievrly) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:271]: OK

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 2 above)
Comment (slievrly) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:293]: At  least, tableMeta information should not be included

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 6]
```diff
@@ -259,6 +267,74 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
...
+        }
+    }
+
+    private static boolean shouldBeRefreshed(ConnectionProxy connectionProxy, String tableName, String columnName) {
+        TableMeta cacheTableMeta = getCacheTableMeta(connectionProxy, tableName);
+        return cacheTableMeta.getColumnMeta(columnName) == null;
+    }
+
+
+    private static TableMeta getCacheTableMeta(ConnectionProxy connectionProxy, String tableName) {
```
Comment (caohdgege) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:332]: TableRecords 这个类是 struct 包下面的，为什么会包含这个方法的，在职能上是不是有点超出范围了 ？

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 7]
```diff
@@ -259,6 +267,74 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
...
+    /**
+     * Build records table records.
+     *
+     * @param tmeta     the tmeta
+     * @param resultSet the result set
+     * @param statementProxy the statement proxy
+     * @return the table records
+     * @throws SQLException the sql exception
+     */
+    public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet, StatementProxy statementProxy)
```
Comment (caohdgege) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:294]: 刷新TableMeta & 重试的这个逻辑放在Executor那一层会不会合理一点 ？

Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:294]: 讨论结论是改成异步刷新了，这一块会移出去

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 6 above)
Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:332]: 讨论结论是改成异步刷新了，这一块会移出去

File: rm-datasource/src/main/java/io/seata/rm/datasource/exception/TableMetaException.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * The type TableMetaException exception.
+ *
+ * @author Bughue
+ */
+public class TableMetaException extends SQLException {
+    private String columnName;
+    private String tableName;
+    private boolean refreshable;
+
+    public TableMetaException(String columnName, String tableName,boolean refreshable) {
```
[the remark this answers is not in the document]
Comment (Bughue) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exception/TableMetaException.java:30]: 抛出的地方好像没看到哪里能拿ResourceId，这个的作用是什么？
其他两个问题我改一下

**Pull request closed without merging**